File: src/codecs/image/icf-cli.ts
#!/usr/bin/env node

/**
 * ICF (Image Codec Format) CLI Tool
 * 
 * Command-line interface for encoding and decoding images using the ICF format.
 */

import * as fs from 'fs';
import * as path from 'path';
import sharp from 'sharp';
import { ImageCodec } from './icf-codec';

function showUsage(): void {
  console.log(`
ICF Image Codec CLI

Usage:
  icf encode <input.jpg|png> <output.icf> [quality]  - Encode image to ICF format
  icf decode <input.icf> <output.jpg|png>            - Decode ICF file to image
  icf info <input.icf>                               - Show ICF file information
  icf help                                           - Show this help message

Parameters:
  quality    - Compression quality (1-100, default: 85)

Examples:
  icf encode photo.jpg photo.icf 90
  icf decode photo.icf photo_decoded.jpg
  icf info photo.icf
`);
}

async function encodeImage(inputPath: string, outputPath: string, quality: number = 85): Promise<void> {
  try {
    console.log(`Encoding ${inputPath} to ${outputPath} (quality: ${quality})...`);
    
    // Check if input file exists
    if (!fs.existsSync(inputPath)) {
      throw new Error(`Input file not found: ${inputPath}`);
    }
    
    // Get original file size
    const originalStats = fs.statSync(inputPath);
    const originalSize = originalStats.size;
    
    // Encode to ICF
    const icfData = await ImageCodec.encode(inputPath, quality);
    
    // Write output file
    fs.writeFileSync(outputPath, icfData);
    
    const compressedSize = icfData.length;
    const compressionRatio = originalSize / compressedSize;
    const savings = ((originalSize - compressedSize) / originalSize) * 100;
    
    console.log(`✓ Encoding completed successfully!`);
    console.log(`  Original size: ${formatBytes(originalSize)}`);
    console.log(`  Compressed size: ${formatBytes(compressedSize)}`);
    console.log(`  Compression ratio: ${compressionRatio.toFixed(2)}:1`);
    console.log(`  Space savings: ${savings.toFixed(1)}%`);
    
  } catch (error: any) {
    console.error(`✗ Encoding failed: ${error.message}`);
    process.exit(1);
  }
}

async function decodeImage(inputPath: string, outputPath: string): Promise<void> {
  try {
    console.log(`Decoding ${inputPath} to ${outputPath}...`);
    
    // Check if input file exists
    if (!fs.existsSync(inputPath)) {
      throw new Error(`Input file not found: ${inputPath}`);
    }
    
    // Read ICF file
    const icfData = fs.readFileSync(inputPath);
    
    // Decode ICF data
    const { data, width, height, channels } = await ImageCodec.decode(icfData);
    
    // Determine output format from extension
    const ext = path.extname(outputPath).toLowerCase();
    let format: 'jpeg' | 'png' | 'webp' = 'jpeg';
    
    switch (ext) {
      case '.png':
        format = 'png';
        break;
      case '.webp':
        format = 'webp';
        break;
      case '.jpg':
      case '.jpeg':
      default:
        format = 'jpeg';
        break;
    }
    
    // Create Sharp image from raw data
    await sharp(data, {
      raw: {
        width,
        height,
        channels: channels as 1 | 2 | 3 | 4
      }
    })
    .toFormat(format)
    .toFile(outputPath);
    
    const outputStats = fs.statSync(outputPath);
    
    console.log(`✓ Decoding completed successfully!`);
    console.log(`  Image dimensions: ${width}x${height}`);
    console.log(`  Channels: ${channels}`);
    console.log(`  Output size: ${formatBytes(outputStats.size)}`);
    
  } catch (error: any) {
    console.error(`✗ Decoding failed: ${error.message}`);
    process.exit(1);
  }
}

async function showFileInfo(inputPath: string): Promise<void> {
  try {
    console.log(`Analyzing ${inputPath}...`);
    
    // Check if input file exists
    if (!fs.existsSync(inputPath)) {
      throw new Error(`Input file not found: ${inputPath}`);
    }
    
    // Read ICF file
    const icfData = fs.readFileSync(inputPath);
    
    // Parse header to get information
    const { header } = parseICFHeader(icfData);
    
    console.log(`\nICF File Information:`);
    console.log(`  Format: ${header.magic} v${header.version}`);
    console.log(`  Dimensions: ${header.width}x${header.height}`);
    console.log(`  Channels: ${header.channels}`);
    console.log(`  Color Space: ${header.colorSpace}`);
    console.log(`  Quality: ${header.quality}%`);
    console.log(`  Compressed Size: ${formatBytes(header.compressedSize)}`);
    console.log(`  File Size: ${formatBytes(icfData.length)}`);
    console.log(`  Checksum: ${header.checksum.substring(0, 16)}...`);
    
    // Calculate estimated original size
    const estimatedOriginalSize = header.width * header.height * header.channels;
    const compressionRatio = estimatedOriginalSize / header.compressedSize;
    
    console.log(`\nCompression Statistics:`);
    console.log(`  Estimated original: ${formatBytes(estimatedOriginalSize)}`);
    console.log(`  Compression ratio: ${compressionRatio.toFixed(2)}:1`);
    
  } catch (error: any) {
    console.error(`✗ Analysis failed: ${error.message}`);
    process.exit(1);
  }
}

function parseICFHeader(icfData: Buffer): { header: any } {
  if (icfData.length < 8) {
    throw new Error('Invalid ICF file: too small');
  }
  
  const magic = icfData.subarray(0, 4).toString('ascii');
  if (magic !== 'ICF1') {
    throw new Error('Invalid ICF magic number');
  }
  
  const headerSize = icfData.readUInt32LE(4);
  if (icfData.length < 8 + headerSize) {
    throw new Error('Invalid ICF file: header size mismatch');
  }
  
  const headerData = icfData.subarray(8, 8 + headerSize);
  const header = JSON.parse(headerData.toString('utf8'));
  
  return { header };
}

function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B';
  
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

function validateQuality(quality: string): number {
  const q = parseInt(quality, 10);
  if (isNaN(q) || q < 1 || q > 100) {
    throw new Error('Quality must be a number between 1 and 100');
  }
  return q;
}

// Main CLI logic
async function main(): Promise<void> {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args[0] === 'help') {
    showUsage();
    return;
  }
  
  const command = args[0];
  
  try {
    switch (command) {
      case 'encode':
        if (args.length < 3) {
          console.error('Error: encode command requires input and output file paths');
          showUsage();
          process.exit(1);
        }
        
        const quality = args[3] ? validateQuality(args[3]) : 85;
        await encodeImage(args[1], args[2], quality);
        break;
        
      case 'decode':
        if (args.length !== 3) {
          console.error('Error: decode command requires input and output file paths');
          showUsage();
          process.exit(1);
        }
        await decodeImage(args[1], args[2]);
        break;
        
      case 'info':
        if (args.length !== 2) {
          console.error('Error: info command requires input file path');
          showUsage();
          process.exit(1);
        }
        await showFileInfo(args[1]);
        break;
        
      default:
        console.error(`Error: Unknown command '${command}'`);
        showUsage();
        process.exit(1);
    }
  } catch (error: any) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

File: src/bin/icf_cli.rs
use clap::{Arg, Command};
use codec_cdn_rust::codecs::image::{IcfCodec, ImageCompressionStats};
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = Command::new("icf-cli")
        .version("1.0")
        .about("Image Codec Format (ICF) CLI tool with advanced DCT compression")
        .subcommand(
            Command::new("encode")
                .about("Encode image to ICF format")
                .arg(
                    Arg::new("input")
                        .help("Input image file (JPEG, PNG, WebP, etc.)")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("output")
                        .help("Output ICF file")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("quality")
                        .help("Quality level (1-100, default: 85)")
                        .short('q')
                        .long("quality")
                        .value_name("NUM")
                        .default_value("85")
                )
        )
        .subcommand(
            Command::new("decode")
                .about("Decode ICF file to image")
                .arg(
                    Arg::new("input")
                        .help("Input ICF file")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("output")
                        .help("Output image file")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .subcommand(
            Command::new("info")
                .about("Show information about ICF file")
                .arg(
                    Arg::new("input")
                        .help("Input ICF file")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .subcommand(
            Command::new("compare")
                .about("Compare original and compressed images")
                .arg(
                    Arg::new("original")
                        .help("Original image file")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("icf")
                        .help("ICF compressed file")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .get_matches();

    let codec = IcfCodec::new();

    match matches.subcommand() {
        Some(("encode", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            let output = sub_matches.get_one::<String>("output").unwrap();
            let quality = sub_matches.get_one::<String>("quality").unwrap()
                .parse::<u8>()
                .map_err(|_| "Quality must be a number between 1 and 100")?;
            
            if quality < 1 || quality > 100 {
                return Err("Quality must be between 1 and 100".into());
            }
            
            println!("Encoding image: {} (quality: {})", input, quality);
            
            let compressed = codec.encode(input, quality)?;
            fs::write(output, &compressed)?;
            
            let stats = codec.get_stats(input, &compressed)?;
            println!("✓ Encoding complete!");
            println!("  Input: {} bytes", stats.original_size);
            println!("  Output: {} bytes", stats.compressed_size);
            println!("  Compression ratio: {:.2}:1", stats.compression_ratio);
            println!("  Space savings: {:.2}%", stats.savings_percent);
        }
        
        Some(("decode", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            let output = sub_matches.get_one::<String>("output").unwrap();
            
            let compressed = fs::read(input)?;
            println!("Decoding ICF file: {} ({} bytes)", input, compressed.len());
            
            let image = codec.decode(&compressed)?;
            image.save(output)?;
            
            println!("✓ Decoding complete!");
            println!("  Output: {} ({}x{})", output, image.width(), image.height());
        }
        
        Some(("info", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            
            let compressed = fs::read(input)?;
            
            // Parse header to show information
            if let Ok((header, _)) = codec.parse_container(&compressed) {
                println!("ICF File Information:");
                println!("  Magic: {}", header.magic);
                println!("  Version: {}", header.version);
                println!("  Dimensions: {}x{}", header.width, header.height);
                println!("  Channels: {}", header.channels);
                println!("  Color space: {}", header.color_space);
                println!("  Quality: {}", header.quality);
                println!("  Compression method: {}", header.compression_method);
                println!("  Block size: {}x{}", header.block_size, header.block_size);
                println!("  Original size: {} bytes", header.original_size);
                println!("  Compressed size: {} bytes", header.compressed_size);
                println!("  File size: {} bytes", compressed.len());
                println!("  Checksum: {}", header.checksum);
                
                let compression_ratio = header.original_size as f64 / compressed.len() as f64;
                let savings = ((header.original_size - compressed.len() as u64) as f64 / header.original_size as f64) * 100.0;
                println!("  Compression ratio: {:.2}:1", compression_ratio);
                println!("  Space savings: {:.2}%", savings);
            } else {
                return Err("Failed to parse ICF header".into());
            }
        }
        
        Some(("compare", sub_matches)) => {
            let original = sub_matches.get_one::<String>("original").unwrap();
            let icf_file = sub_matches.get_one::<String>("icf").unwrap();
            
            let compressed = fs::read(icf_file)?;
            let stats = codec.get_stats(original, &compressed)?;
            
            println!("Image Compression Comparison:");
            println!("  Original: {}", original);
            println!("  Compressed: {}", icf_file);
            println!("  {}", stats);
            
            // Decode and compare dimensions
            let decoded = codec.decode(&compressed)?;
            let original_img = image::open(original)?;
            
            println!("  Original dimensions: {}x{}", original_img.width(), original_img.height());
            println!("  Decoded dimensions: {}x{}", decoded.width(), decoded.height());
            
            if original_img.width() == decoded.width() && original_img.height() == decoded.height() {
                println!("  ✓ Dimensions match");
            } else {
                println!("  ⚠ Dimension mismatch");
            }
        }
        
        _ => {
            eprintln!("No subcommand provided. Use --help for usage information.");
            std::process::exit(1);
        }
    }
    
    Ok(())
}

// Usage examples:
// icf-cli encode input.jpg output.icf --quality 85
// icf-cli decode output.icf decoded.png
// icf-cli info output.icf
// icf-cli compare input.jpg output.icf